onnxruntime/core/optimizer/matmul_add_fusion.cc

```diff
@@ -69,25 +74,32 @@ Status MatMulAddFusion::ApplyImpl(Graph& graph, bool& modified, int graph_level,
       continue;
     }
 
-    auto matmul_output_name = matmul_node.OutputDefs()[0]->Name();
+    const auto& matmul_output = *matmul_node.OutputDefs()[0];
+
+    auto matmul_output_name = matmul_output.Name();
     auto gemm_input_defs = matmul_input_defs;
     if (matmul_output_name == add_input_defs[0]->Name()) {
       // matmul output as Add_A, should use Add_B as input C for gemm
-      // Gemm only support unidirectional broadcast on C
-      if (add_input_defs[1]->Shape()->dim_size() > 2) {
-        continue;
-      }
       gemm_input_defs.push_back(add_input_defs[1]);
     } else {
       // matmul output as Add_B, should use Add_A as input C for gemm
-      // Gemm only support unidirectional broadcast on C
-      if (add_input_defs[0]->Shape()->dim_size() > 2) {
-        continue;
-      }
       gemm_input_defs.push_back(add_input_defs[0]);
     }
 
-    if (!graph.GetNodeOutputsInGraphOutputs(node).empty()) {
+    // valid bias_shapes are (N) or (1, N) or (M, 1) or (M, N) as
+    // GEMM only supports unidirectional broadcast on the bias input C
+    const auto& bias_shape = *gemm_input_defs.back()->Shape();
+    const auto& M = matmul_output.Shape()->dim()[0];
+    const auto& N = matmul_output.Shape()->dim()[1];
```

Comment:
@skottmckay ,thanks for the quick PR!  

For the dim()[1] check above, just wondering if it's always safe? From the doc below, it seems like if we multiply  2 vectors of shape (1,N) * (N,1), we should end up with a scalar, which I'm guessing may not be stored with 2-D shape. Not sure what's our exact implementation of MatMul, but might be good idea to check in case dim()[1] does not exist.

https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.matmul.html

Reply:
It think it should be safe based on earlier checks in this method (line 72) that ensure both inputs are rank 2:

```
    if (2 != matmul_a_shape->dim_size() || 2 != matmul_b_shape->dim_size()) {
      // Gemm only support Matrix
      continue;
    }
```

and the MatMul ONNX shape inferencing where it uses dim 0 from the first input and dim 1 from the second when each is rank 2.

Reply:
Ok, great - line 72 was just above the fold